### PR TITLE
helm: Add conditional creation of serviceAccount

### LINKED ISSUE #11866: Add conditional creation of teleport serviceaccount in helm chart
### What
We use [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) helm chart to manage teleport HA deployment. We manage our service accounts separately, and so we don't want the teleport chart to create its own Service account. 
### How
We should have the ability to enable/disable service account creation in `values.yaml`. If the service account is disabled we should also be able to pass the name of the service account we would like to use.
Looking through the code base, this feature was implemented in the [deprecated teleport chart](https://github.com/gravitational/teleport/blob/master/examples/chart/teleport/templates/serviceaccount.yaml). 
Something similar to this:
Values configuration.
```
serviceAccount:
  # Specifies whether a ServiceAccount should be created
  create: true
  # The name of the ServiceAccount to use.
  # If not set and create is true, a name is generated using the fullname template
  name:
  annotations: {}
```

And in service account templates.
```
{{- if .Values.serviceAccount.create -}}
apiVersion: v1
kind: ServiceAccount
metadata:
  name: {{ template "teleport.serviceAccountName" . }}
  labels:
  {{- include "teleport.labels" . | nindent 4 }}
  {{- if .Values.serviceAccount.annotations }}
  annotations: {{ toYaml .Values.serviceAccount.annotations | nindent 4 }}
  {{- end }}
{{- end }}

```

### Why
We spin up eks clusters dynamically and for all components that require AWS access, we crate service accounts and AWS roles with cross-plane. This means that the chart who are going to access AWS must-have service account creation is disabled, otherwise, they override the `eks.amazonaws.com/role-arn:` annotation.

### Workaround
We can add `eks.amazonaws.com/role-arn:` manually in the values.yaml. But if we change policy/role in our crossplane configuration, we will require to manually change arn in all teleport values files.

### PR DESCRIPTION
**Pull request by nickolanick**
This pr resolves https://github.com/gravitational/teleport/issues/11866
### WHAT
- conditionally create the service account.
- allow users to pass the name of an existing service account.
### Notes
Changes in this pr are back portable, they will not break any current usage of this chart.
if `serviceAccount:` is not specified the behavior does not change, and all tests are successful

### CODE REVIEW

File: examples/chart/teleport-cluster/templates/serviceaccount.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/templates/serviceaccount.yaml:1]:
```suggestion
{{- if .Values.serviceAccount.create -}}
```

File: examples/chart/teleport-cluster/values.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/values.yaml:245]:
```suggestion
  # To set annotations on the service account, use the annotations.serviceAccount value.
```

I don't like the idea of having annotations set in two places. Please remove this and replace with a comment telling people to use `annotations.serviceAccount` instead.

File: examples/chart/teleport-cluster/values.schema.json
Comment (webvictim) [inline examples/chart/teleport-cluster/values.schema.json:471]:
```suggestion
                }
```

File: examples/chart/teleport-cluster/templates/_helpers.tpl
Comment (webvictim) [inline examples/chart/teleport-cluster/templates/_helpers.tpl:10]:
```suggestion
  {{- coalesce .Values.serviceAccount.name .Release.Name -}}
```

File: examples/chart/teleport-cluster/.lint/service-account.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/.lint/service-account.yaml:6]:
```suggestion
annotations:
  serviceAccount: "test-annotation"
```

File: examples/chart/teleport-cluster/templates/_helpers.tpl
Comment (webvictim) [inline examples/chart/teleport-cluster/templates/_helpers.tpl:3]:
```suggestion
if serviceAccount is not defined or serviceAccount.name is empty, use .Release.Name
```

File: examples/chart/teleport-cluster/templates/clusterrolebinding.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/templates/clusterrolebinding.yaml:11]:
```suggestion
  name: {{ template "teleport.serviceAccountName" . }}
```

File: examples/chart/teleport-cluster/templates/serviceaccount.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/templates/serviceaccount.yaml:11]:
```suggestion
{{- if .Values.annotations.serviceAccount }}
  annotations:
{{- toYaml .Values.annotations.serviceAccount | nindent 4 }}
{{- end -}} 
```

File: examples/chart/teleport-cluster/values.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/values.yaml:243]:
```suggestion
  name: ""
```

File: examples/chart/teleport-cluster/values.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/values.yaml:241]:
```suggestion
  # If not set and serviceAccount.create is true, the name is generated using the release name.
```

File: examples/chart/teleport-cluster/values.yaml
Comment (webvictim) [inline examples/chart/teleport-cluster/values.yaml:242]:
```suggestion
  # If create is false, the name will be used to reference an existing service account.
```

**Review by webvictim — CHANGES_REQUESTED** [review]
Thanks for the PR. I've requested some changes as I think there's a cleaner way to achieve this.

Please also take a look at our Helm chart contribution guidelines: https://github.com/gravitational/teleport/blob/master/examples/chart/CONTRIBUTING.md

These new values would also need adding to the `teleport-cluster` chart reference: https://github.com/gravitational/teleport/blob/master/docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx

File: examples/chart/teleport-cluster/templates/serviceaccount.yaml
Comment (nickolanick) [reply examples/chart/teleport-cluster/templates/serviceaccount.yaml]: This is a potentially breakable change, as after chart upgrade, for those who will use old values, this will raise a nullable error, as the `serviceAccount` field will not be defined. If that's not a concern, then all is good.

**Comment by nickolanick** [comment]
@webvictim Thanks for the comments, resolved.

**Comment by nickolanick** [comment]
@webvictim Any updates on this one?

**Comment by webvictim** [comment]
@nickolanick Lint is failing:

```
Using values from 'examples/chart/teleport-cluster/.lint/service-account.yaml': examples/chart/teleport-cluster/.lint/service-account.yaml
  6:18      error    trailing spaces  (trailing-spaces)

     1	clusterName: helm-lint$
     2	serviceAccount:$
     3	  create: true$
     4	  name: helm-lint$
     5	annotations:$
     6	  serviceAccount: $
     7	    kubernetes.io/serviceaccount: "test-annotation"$
Makefile:702: recipe for target 'lint-helm' failed
```

**Comment by zmb3** [comment]
@webvictim @hugoShaka is there still an appetite to merge this?

**Comment by hugoShaka** [comment]
I think we can proceed with this. I'm curious about the use-case (you still need rights to create the RoleBinding, so why do you need to create it beforehand ?) but `serviceAccount.create` seems like a generic value used in common charts (I checked cert-manager, external-dns and prometheus-kube-stack). Users can legitimately expect our chart to offer the same parameter than the rest of the ecosystem.

**Comment by webvictim** [comment]
I agree with Hugo.

**Comment by nickolanick** [comment]
For some reason, I thought that [#10515](https://github.com/gravitational/teleport/pull/10515) was going to be merged instead of this one, and never finished it ... @webvictim Is there anything that still needs to be done here? Would be happy to resolve and merge this.

@hugoShaka As for use cases. There are often situations when you want to create a service account outside of the helm chart; for example, we create a service account with `IRSA annotations` from terraform/crossplane and don't want the teleport chart to overwrite it.

**Review by hugoShaka — APPROVED** [review]
LGTM

**Review by webvictim — APPROVED** [review]

**Comment by hugoShaka** [comment]
Hi @nickolanick ,

Thanks for your contribution. We are currently unable to run our tests on external contributions. As such, I'll be helping you validate and merge your pull request. I've created an internal PR here:

https://github.com/gravitational/teleport/pull/16337

Please wait for another approver and I'll get your code merged.

**Pull request closed without merging**
